### ISSUE TITLE
Set up monitoring and alerting for Via 3

### ISSUE THREAD

**Issue by seanh** [hypothesis/via#10]
Via 3 needs to have all the usual monitoring and alerting--the same as h and LMS have--before we can enable it for LMS in production.

Some of these may already be set up, in which case just check them off.

The via3.hypothes.is service should have:

- [x] Dependabot.

  This is working correctly for Docker and Python -- seanh

- [x] Papertrail logging for QA and prod

  Logs from both the Python code and from Nginx should be being captured in Papertrail.

  This is **needed for both QA and prod**.

  Must include:

  - [x] All logs from all processes running in the Docker container (which is just Gunicorn and NGINX I think) going to Papertrail
  - [ ] I don't believe we need/want this: All the other non-Docker logs in the VM's `/var/logs/` 

- [x] Sentry crash reporting for QA and prod, Python.

  This is working correctly. It's Python only, not JavaScript --seanh.

- [x] New Relic for prod.

  This didn't appear to be set up quite right.

  @dmfine Can you also review what alerts are set in New Relic?
  We want to monitor both the performance of the <del>Nginx and</del> Python parts of
  py-proxy using New Relic. It might be good to also set up some New Relic alerts
  that fire if performance dips too low.

- [x] Pingdom

  This was not set up. Also, Via 2 needs to be removed from Pingdom.

- [x] PagerDuty.

  Unsure whether this is set up correctly.

  If Via 3 goes down a PagerDuty alert should fire. Via 3 should be treated as
  a critical web service, the same as h and LMS.

- [x] Add some sensible alerts in Elastic Beanstalk

  We don't need EB alerts for things that NR can already alert on (response status, response time).

  There are a couple of alerts in here that look like they just duplicate things that NR already alerts on. Remove these.

  If possible we would like some infrastructure-based EB alerts, e.g. system load, CPU usage, memory usage, saturation of IO upload or download bandwidth. Or generally anything helpful that EB can do that NR can't

  There may not be much that EB can do for us here, in which case never mind

- [x] Monitoring of **queueing**. And alerting based on.

  This can probably be done in New Relic (?). We would like to monitor how much request queuing is happening and alert on too much queuing

**Comment by hmstepanek** [comment]
Hi @seanh there seems to be a slight misunderstanding here-we have to pay for nginx monitoring inside New Relic and we currently aren't doing this as we decided the nginx monitoring wasn't really interesting so nginx-level monitoring is purely in AWS/logs.

**Comment by seanh** [comment]
How much will it cost to enable Nginx monitoring? I think that because proxying things through Nginx is sort of the main thing that this new app does, it will be worth paying to have Nginx monitoring in the same place as where we have the rest of our performance monitoring, and to have the better monitoring UI and features (e.g. alerts etc) of New Relic.

Presumably if we pay for Nginx monitoring for Via 3 we'll also then get Nginx monitoring in New Relic for the rest of our apps too, which will be a bonus.

**Comment by hmstepanek** [comment]
Just making a note of what @seanh and I discussed earlier this am in regards to the comment above. When we looked at the nginx new relic monitoring it was simply querying the nginx api and gave us some basic stats like response codes and request time. These two things are available in aws for free so there's really no sense in paying for them. That being said nginx specific and infrastructure monitoring in general is a separate issue that we don't want to get hung up on here. We should make a separate card and look into this later if we find it to be necessary.

**Comment by hmstepanek** [comment]
I'm not going to add any sentry crash reporting for frontend js right now. The frontend is mainly pdf js with some html/simple js that we've added to it to configure the client. This is the same code that is in both via apps in production right now and we don't have frontend crash reporting for those either. We can always add it later but I don't see a need for it at the moment.

**Comment by hmstepanek** [comment]
I've setup a warning to output in slack when the response time is >3.4s for 5 mins. I'm basing this off the incident we had today in via where the response time was between 5s and 2s during the time of the incident. If this works well we can turn it into a pager duty alert but I don't want to do that immediately in case it needs to be adjusted.

**Comment by dmfine** [comment]
Editing to remove QA requirement for New Relic. I don't think we want automated alerts for anything other than prod, we should be looking at QA metrics manually when pertinent. Regarding papertrail, we don't collect /var/log on the host for any of our other apps, I think of papertrail specifically for application logs. If we want to change it, we should change it across all our apps. Which isn't that hard but let's sync on that first.

**Comment by dmfine** [comment]
I setup one elastic beanstalk alert on "environment healthy". I created an NR alert on Queueing but I'm not certain it is measuring what we want. Will discuss tomorrow.

**Comment by dmfine** [comment]
It looks like the queuing alert is capturing what we want. I have a PR open for the final change to papertrail logs.

**Comment by seanh** [comment]
Re-opening this since the PRs to fix Papertrail haven't been reviewed, merged, deployed or tested yet

**Comment by seanh** [comment]
I've changed both Via 3 QA and prod to send all logs from `/var/log/*` and `/var/log/**/*` to Papertrail:

https://github.com/hypothesis/deployment/blob/4e957c96434ee33513b0bd9affa73e2627f5374b/via3/ebextensions/qa/papertrail.config#L20-L22

https://github.com/hypothesis/deployment/blob/589a09a4374fc124b033281001027907babd1155/via3/ebextensions/prod/papertrail.config#L20-L22

And done a sync-env and re-deploy (for good measure) of both QA and prod.

<del>Nonetheless:</del>

* <del>I'm only seeing `access.log` and `eb-*-stdouterr.log` in Papertrail, not the rest of the files</del>

* <del>If I ssh into a Via 3 QA or prod app server and look at the `/etc/log_files.yml` it contains the old config, where it's configured to capture only Docker and NGINX logs, not `/var/log/*`.</del>

<del>So it looks as if the changes to Via 3 ebextensions files in the deployment repo are not being deployed to QA or prod by a sync-env or re-deploy.</del>

Update: fixed this. You actually have to _deploy_ an app for ebextensions changes to take effect, an env-sync or redeploy doesn't work. I'll post an SO4T answer

**Comment by seanh** [comment]
Now that we're capturing _all_ the logs from `/var/logs/`, we're capturing some regular log messages that aren't interesting. It's not a large volume of them but there are a few regular ones (every second or so). I'm experimenting with filtering these out using regexes in Papertrail. On a one-regex-per-message basis for now, though we could get more expansive in future if needed.

We can also omit entire log files from the Papertrail config in the first place, but it's probably safer and easier to just filter out individual messages.

They recommend using https://rubular.com/ to test your regexes (against actually log lines copied from the logs).

![Screenshot from 2020-02-06 19-44-16](https://user-images.githubusercontent.com/22498/73972803-5bb7be80-4919-11ea-8a04-65f7824377eb.png)
